File: dashboard/src/service/api/backup.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { $fetch, fetcher } from '@/service/http'

export interface BackupRemoteSftp {
  enabled: boolean
  host: string
  port: number
  username: string
  remote_path: string
}

export interface BackupConfig {
  auto_enabled: boolean
  schedule_hours: number
  local_retention: number
  upload_to_remote: boolean
  remote: BackupRemoteSftp
}

export interface BackupListItem {
  id: string
  created_at: string
  panel_version: string
  database_engine: string
  size_bytes: number
  sha256: string
  encrypted?: boolean
  remote_uploaded: boolean
  filename: string
}

export interface BackupListResponse {
  items: BackupListItem[]
  status: 'idle' | 'running' | 'success' | 'failed'
  last_error: string
  last_success_at?: string | null
  config: BackupConfig
}

export const getBackups = () => fetcher<BackupListResponse>('/api/backup')

export const updateBackupConfig = (config: BackupConfig) =>
  fetcher<BackupConfig>('/api/backup/config', { method: 'PUT', body: config })

export const runBackup = () => fetcher<{ manifest: { id: string }; message: string }>('/api/backup/run', { method: 'POST' })

export const restoreBackup = (backupId: string, dryRun = false) =>
  fetcher<{ success: boolean; message: string; restart_required: boolean; dry_run?: boolean; checks?: string[] }>(
    `/api/backup/${backupId}/restore?dry_run=${dryRun ? 'true' : 'false'}`,
    {
      method: 'POST',
      // Restore can take a few minutes; never hang the UI forever if the API stalls.
      timeout: dryRun ? 60_000 : 360_000,
    },
  )

export const validateBackup = (backupId: string) => restoreBackup(backupId, true)

export interface BackupInstallTokenResponse {
  token: string
  backup_id: string
  filename: string
  expires_at: string
  restore_url: string
  install_command: string
  note?: string
}

export interface BackupInstallTokenListItem {
  token: string
  backup_id: string
  filename: string
  created_at?: string | null
  expires_at?: string | null
  enabled: boolean
  revoked: boolean
  consumed: boolean
  use_count: number
  last_used_at?: string | null
  last_used_ip?: string | null
  last_used_country?: string | null
  last_used_country_code?: string | null
  last_used_city?: string | null
  last_used_isp?: string | null
  last_used_asn?: string | null
}

export const createBackupInstallToken = (
  backupId: string,
  body?: { panel_base_url?: string; database?: string; ttl_hours?: number },
) =>
  fetcher<BackupInstallTokenResponse>(`/api/backup/${backupId}/install-token`, {
    method: 'POST',
    body: body ?? { database: 'timescaledb', ttl_hours: 72 },
  })

export const getBackupInstallTokens = () =>
  fetcher<{ items: BackupInstallTokenListItem[] }>('/api/backup/install-tokens')

export const updateBackupInstallToken = (token: string, enabled: boolean) =>
  fetcher(`/api/backup/install-tokens/${encodeURIComponent(token)}`, {
    method: 'PATCH',
    body: { enabled },
  })

export const revokeBackupInstallToken = (token: string) =>
  fetcher(`/api/backup/install-tokens/${encodeURIComponent(token)}/revoke`, { method: 'POST' })

export const importBackupArchive = async (file: File) => {
  const form = new FormData()
  form.append('file', file)
  return $fetch<{ manifest: { id: string }; message: string }>('/api/backup/import', { method: 'POST', body: form })
}

export const downloadBackupUrl = (backupId: string) => `/api/backup/${backupId}/download`

export const downloadBackupFile = async (backupId: string, filename: string) => {
  const blob = await $fetch<Blob>(downloadBackupUrl(backupId), { responseType: 'blob' })
  const url = window.URL.createObjectURL(blob)
  const anchor = document.createElement('a')
  anchor.href = url
  anchor.download = filename
  document.body.appendChild(anchor)
  anchor.click()
  document.body.removeChild(anchor)
  window.URL.revokeObjectURL(url)
}

export const useBackups = () =>
  useQuery({
    queryKey: ['backup', 'list'],
    queryFn: getBackups,
    refetchInterval: 5000,
  })

export const useUpdateBackupConfig = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: updateBackupConfig,
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['backup', 'list'] }),
  })
}

export const useRunBackup = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: runBackup,
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['backup', 'list'] }),
  })
}

export const useRestoreBackup = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (backupId: string) => restoreBackup(backupId, false),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['backup', 'list'] }),
  })
}

export const useValidateBackup = () =>
  useMutation({
    mutationFn: (backupId: string) => validateBackup(backupId),
  })

export const useCreateBackupInstallToken = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (backupId: string) => createBackupInstallToken(backupId),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['backup', 'install-tokens'] }),
  })
}

export const useBackupInstallTokens = (enabled = true) =>
  useQuery({
    queryKey: ['backup', 'install-tokens'],
    queryFn: getBackupInstallTokens,
    enabled,
    refetchInterval: 5000,
  })

export const useUpdateBackupInstallToken = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ token, enabled }: { token: string; enabled: boolean }) => updateBackupInstallToken(token, enabled),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['backup', 'install-tokens'] }),
  })
}

export const useRevokeBackupInstallToken = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (token: string) => revokeBackupInstallToken(token),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['backup', 'install-tokens'] }),
  })
}

File: app/models/backup.py
from datetime import datetime as dt
from enum import StrEnum

from pydantic import BaseModel, Field


class BackupStatus(StrEnum):
    idle = "idle"
    running = "running"
    success = "success"
    failed = "failed"


class BackupRemoteSftp(BaseModel):
    enabled: bool = False
    host: str = ""
    port: int = Field(default=22, ge=1, le=65535)
    username: str = ""
    remote_path: str = "/var/backups/hpxpanel"


class BackupConfig(BaseModel):
    auto_enabled: bool = False
    schedule_hours: int = Field(default=24, ge=1, le=168)
    local_retention: int = Field(default=14, ge=1, le=365)
    upload_to_remote: bool = True
    remote: BackupRemoteSftp = Field(default_factory=BackupRemoteSftp)


class BackupManifest(BaseModel):
    id: str
    created_at: dt
    panel_version: str
    database_engine: str
    database_file: str
    size_bytes: int
    sha256: str
    encrypted: bool = False
    remote_uploaded: bool = False
    remote_error: str = ""


class BackupListItem(BaseModel):
    id: str
    created_at: dt
    panel_version: str
    database_engine: str
    size_bytes: int
    sha256: str
    encrypted: bool = False
    remote_uploaded: bool = False
    filename: str


class BackupListResponse(BaseModel):
    items: list[BackupListItem] = Field(default_factory=list)
    status: BackupStatus = BackupStatus.idle
    last_error: str = ""
    last_success_at: dt | None = None
    config: BackupConfig


class BackupRunResponse(BaseModel):
    manifest: BackupManifest
    message: str = ""


class BackupRestoreResponse(BaseModel):
    success: bool
    message: str
    restart_required: bool = True
    dry_run: bool = False
    checks: list[str] = Field(default_factory=list)


class BackupInstallTokenRequest(BaseModel):
    panel_base_url: str | None = None
    database: str = "timescaledb"
    ttl_hours: int = Field(default=72, ge=1, le=168)


class BackupInstallTokenResponse(BaseModel):
    token: str
    backup_id: str
    filename: str
    expires_at: dt
    restore_url: str
    install_command: str
    note: str = (
        "Keep the old panel online until the new server finishes downloading. "
        "Token expires automatically; treat it like a password."
    )


class BackupInstallTokenListItem(BaseModel):
    token: str
    backup_id: str
    filename: str
    created_at: dt | None = None
    expires_at: dt | None = None
    enabled: bool = True
    revoked: bool = False
    consumed: bool = False
    use_count: int = 0
    last_used_at: dt | None = None
    last_used_ip: str | None = None
    last_used_country: str | None = None
    last_used_country_code: str | None = None
    last_used_city: str | None = None
    last_used_isp: str | None = None
    last_used_asn: str | None = None


class BackupInstallTokensResponse(BaseModel):
    items: list[BackupInstallTokenListItem] = Field(default_factory=list)


class BackupInstallTokenUpdate(BaseModel):
    enabled: bool

